### ISSUE TITLE
HttpClient does not always return 'Authorization' header correctly

### ISSUE THREAD

**Issue by dustinsoftware** [dotnet/corefx#31143]
Simple app to repro this behavior is in [this gist](https://gist.github.com/dustinsoftware/424ba34074f4044a1420713bda37104a), on .NET Core 2.1.302.

When accessing the `Authorization` header in .NET Core 2.1 via `GetValues`, it will be empty:
```cs
response.Headers.TryGetValues("Authorization", out var authorization);
Console.WriteLine(authorization?.First()); // empty string
```

Calling this method inside netstandard 2.0 library running on .NET 4.7 works as expected.

This always works:
```cs
Console.WriteLine(response.Headers.First(x => x.Key == "Authorization").Value.FirstOrDefault());
```

Is this behavior documented somewhere? Should I be iterating over `Headers` instead?

**Comment by karelz** [comment]
Can you check it against .NET Core 2.0? I wonder if it is recent regression, or "just" difference against .NET Framework

**Comment by dustinsoftware** [comment]
I can't repro with .NET Core 2.0. Output from my [test app](https://gist.github.com/dustinsoftware/424ba34074f4044a1420713bda37104a):

```
.NET Core 4.6.00001.0
System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Headers.TryGetValues(Server): Kestrel
Headers.TryGetValues(Authorization): Bearer dpmtest
Headers.First(x => x.Key == Authorization): Bearer dpmtest
```

**Comment by davidsh** [comment]
cc: @stephentoub 

Were there any changes in .NET Core 2.1 regarding collection classes?  It sounds like this change is due to something in that area and directly the HttpHeaders class.

**Comment by dustinsoftware** [comment]
I’m dealing with a private auth API which sets an Authorization header on
the response (instead of putting the token in the response body). I haven’t
found a public API which I could use as an example, which is why I provided
an example controller action.

**Comment by davidsh** [comment]
Thanks for the added details.

This definitely looks like a regression .NET Core 2.0 to 2.1.  the `TryGetValues()` method should always be able to return the string values of a header given the headername as a string parameter to TryGetValues().

I suspect this broke during some code optimization perhaps with this commit or some other related ones last year:
https://github.com/dotnet/corefx/commit/b963747c6ef3df6a9512854714fcb11b38f1c0c8

cc: @geoffkizer @stephentoub

**Comment by bgrainger** [comment]
> I haven’t found a public API which I could use as an example

I have encountered this same problem with a public API (but it requires authentication, so I can't provide a code sample). GCS resumable uploads set the `Range` header (typically considered a request header) on the response when an upload is interrupted. From https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#resume-upload:

> If you received a `308 Resume Incomplete` response, process the response's `Range` header, which specifies which bytes the server has received so far. The response will not have a `Range` header if no bytes have been received yet.
> 
> For example, a `Range` header of `bytes=0-42` indicates that the first 43 bytes of the file have been received.

I attempted to use `response.Headers.GetValues("Range")` in my `netcoreapp2.1` application, but it threw an `InvalidOperationException`.

**Comment by karelz** [comment]
If it is 2.0->2.1 regression, we should consider it for servicing - setting milestone.

**Comment by dustinsoftware** [comment]
I cloned `corefx` and tried to repro the issue, however I was unable to do so on commit 5bff639f1afd949852bc098fed27c002d5916036. I'll attempt to checkout the 2.1 branch and see if it repro's there...

```csharp
        [Fact]
        public async Task TestAuthHeadersOnResponse()
        {
            using (var client = new HttpClient(new CustomResponseHandler(
                (r, c) =>
                {
                    var result = new HttpResponseMessage(HttpStatusCode.OK)
                    {
                        Content = new ByteArrayContent(new byte[1]),
                    };
                    result.Headers.Add("Authorization", "Bearer dpmtest");
                    result.Headers.Add("X-Test", "dpmtest");

                    return Task.FromResult(result);
                })))
            {
                var headers = await client.GetAsync(CreateFakeUri());
                Assert.Equal("Bearer dpmtest", headers.Headers.GetValues("Authorization").First());
                Assert.Equal("dpmtest", headers.Headers.GetValues("X-Test").First());
            }
        }
```

**Comment by dustinsoftware** [comment]
Following up on this - I did repro on 5bff639 after all, the test example pasted above was not enough to repro the bug. I had to resort to making a real HTTP request to see the unexpected behavior.

**Comment by davidsh** [comment]
### Shiproom template

#### Description
TBD

#### Customer Impact
TBD

#### Regression?
Regression from 2.0.

#### Risk
TBD

#### Link to release/2.1.x PR
TBD

**Comment by karelz** [comment]
Fixed in PR #31768 in release/2.1 branch. It will be part of 2.1.5 release.

**Comment by dustinsoftware** [comment]
Thanks!!
